File: 2086D.cpp
#include <bits/stdc++.h>
using namespace std;
#define int long long
#define rep(i,a,b) for (int i = (a); i < (b); i++)
typedef vector<int> vi;
const int MOD = 998244353;



int solve(int i,int cur, int x, vi &a,vector<vector<int>> &dp) {
    if(cur>x) return 0;
    if(i==a.size()) return cur==x;
    if(dp[i][cur]!=-1) return dp[i][cur];
    int ans = solve(i+1,cur,x,a,dp);
    if(cur+a[i]<=x){
        ans += solve(i+1,cur+a[i],x,a,dp);
        ans %=MOD;
    }
    return dp[i][cur]=ans;
}

const int N = 3*(1e5) + 5;
long long fact[N];
long long inv[N];
long long finv[N];

void pre() {
    fact[0] = 1;
    for (int i = 1; i <= N; i++) {
        fact[i] = (fact[i-1]*i)%MOD;
    }
    inv[1] = 1;
    for (int i = 2; i <= N; i++) {
        inv[i] = 1LL * (MOD - MOD / i) * inv[MOD % i] % MOD;
    }
    finv[0] = finv[1] = 1;
    for (int i = 2; i <= N; i++) {
        finv[i] = (inv[i]*finv[i-1])%MOD;
    }
}

long long nCk(int n,int k) {
    return (((fact[n]*finv[k])%MOD)*finv[n-k])%MOD;
}


void testcase() {
    int S = 0;
    int n = 26;
    vi c(n);
    vi a;
    rep(i, 0, n) cin >> c[i];
    
    rep(i, 0, n) {
        if (c[i]) {
            a.push_back(c[i]);
        }
    }
    
    rep(i, 0, a.size()) {
        S += a[i];
    }
    int target = S / 2;
    vector<vector<int>> dp(a.size(),vector<int>(target+1,-1));
    int cnt = solve(0,0,target,a,dp);
    int ans = cnt;
    ans = (ans * fact[target] % MOD * fact[S-target]%MOD) % MOD;
    int den = 1;
    rep(i, 0, a.size()) {
        den = (den * finv[a[i]]) % MOD;
    }
    ans = (ans * den) % MOD;
    cout << ans << "\n";
}

int32_t main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    pre();
    int T = 1;
    cin >> T;
    while (T--) testcase();
    return 0;
}

File: 2086D.rs
use std::io::{self, BufRead, BufReader, Write};

const MOD: i64 = 998244353;
const N: usize = 14 * 5 * 100_000 + 5; // 7_000_005

fn precompute() -> (Vec<i64>, Vec<i64>, Vec<i64>) {
    let mut fact = vec![0i64; N];
    let mut inv = vec![0i64; N];
    let mut finv = vec![0i64; N];
    fact[0] = 1;
    for i in 1..N {
        fact[i] = fact[i - 1] * (i as i64) % MOD;
    }
    inv[1] = 1;
    for i in 2..N {
        let i_val = i as i64;
        inv[i] = (MOD - MOD / i_val) * inv[(MOD % i_val) as usize] % MOD;
    }
    finv[0] = 1;
    finv[1] = 1;
    for i in 2..N {
        finv[i] = inv[i] * finv[i - 1] % MOD;
    }
    (fact, inv, finv)
}

fn get_subset_sums(a: &[i64]) -> Vec<i64> {
    let len = a.len();
    let total = 1 << len;
    let mut res = Vec::with_capacity(total);
    for mask in 0..total {
        let mut sum = 0;
        let mut m = mask;
        while m != 0 {
            let j = m.trailing_zeros() as usize;
            sum += a[j];
            m &= m - 1;
        }
        res.push(sum);
    }
    res
}
fn solve(a: &[i64], target: i64) -> i64 {
    let n = a.len();
    let mid = n / 2;
    let left = get_subset_sums(&a[..mid]);
    let right = get_subset_sums(&a[mid..]);

    let mut left_sorted = left;
    let mut right_sorted = right;
    left_sorted.sort();
    right_sorted.sort();

    let mut ans = 0;
    for &l_sum in left_sorted.iter() {
        let req = target - l_sum;
        let low = right_sorted.partition_point(|&y| y < req);
        let high = right_sorted.partition_point(|&y| y <= req);
        ans += (high - low) as i64;
    }
    ans
}

fn main() {
    let (fact, _inv, finv) = precompute();
    let stdin = io::stdin();
    let mut reader = BufReader::new(stdin.lock());
    let mut t_line = String::new();
    reader.read_line(&mut t_line).unwrap();
    let t: usize = t_line.trim().parse().unwrap();

    let stdout = io::stdout();
    let mut out = io::BufWriter::new(stdout.lock());

    for _ in 0..t {
        let mut input_line = String::new();
        reader.read_line(&mut input_line).unwrap();
        let c: Vec<i64> = input_line
            .split_whitespace()
            .map(|s| s.parse().unwrap())
            .collect();

        let mut a = Vec::new();
        for &x in c.iter().take(26) {
            if x != 0 {
                a.push(x);
            }
        }
        
        let s_total: i64 = a.iter().sum();
        let target = s_total / 2;
        let cnt = solve(&a, target);
        let mut ans = cnt % MOD;
        ans = ans * fact[target as usize] % MOD * fact[(s_total - target) as usize] % MOD;
        let mut den = 1;
        for &x in &a {
            den = den * finv[x as usize] % MOD;
        }
        ans = ans * den % MOD;
        writeln!(out, "{}", ans).unwrap();
    }
}
